amcl/src/amcl/amcl.cpp

```diff
@@ -0,0 +1,826 @@
+  pf_init_ = false;
+}
+
+void AmclNode::handleMapMessage(const nav_msgs::OccupancyGrid& msg)
+{
+  boost::recursive_mutex::scoped_lock cfl(configuration_mutex_);
+
+  ROS_INFO("Received a %d X %d map @ %.3f m/pix\n",
+           msg.info.width,
+           msg.info.height,
+           msg.info.resolution);
+
+  freeMapDependentMemory();
+  // Clear queued laser objects because they hold pointers to the existing
+  // map, #5202.
+  lasers_.clear();
+  lasers_update_.clear();
+  frame_to_laser_.clear();
+
+  map_ = convertMap(msg);
+
+#if NEW_UNIFORM_SAMPLING
+  // Index of free space
+  free_space_indices.resize(0);
+  for(int i = 0; i < map_->size_x; i++)
+    for(int j = 0; j < map_->size_y; j++)
+      if(map_->cells[MAP_INDEX(map_,i,j)].occ_state == -1)
+        free_space_indices.push_back(std::make_pair(i,j));
+#endif
+  // Create the particle filter
+  pf_vector_t pf_init_pose_mean = pf_vector_zero();
+  pf_init_pose_mean.v[0] = init_pose_[0];
+  pf_init_pose_mean.v[1] = init_pose_[1];
+  pf_init_pose_mean.v[2] = init_pose_[2];
+  pf_matrix_t pf_init_pose_cov = pf_matrix_zero();
+  pf_init_pose_cov.m[0][0] = init_cov_[0];
+  pf_init_pose_cov.m[1][1] = init_cov_[1];
+  pf_init_pose_cov.m[2][2] = init_cov_[2];
+  instantiateParticleFilter(pf_init_pose_mean, pf_init_pose_cov);
+
+  // Instantiate the sensor objects
+  // Odometry
+  instantiateOdom();
+  // Laser
+  instantiateLaser();
+  // In case the initial pose message arrived before the first map,
+  // try to apply the initial pose now that the map has arrived.
+  applyInitialPose();
+}
+
+
+bool AmclNode::handleInitialPosesMessage(const move_base_msgs::PoseWithCovarianceStampedArray& msg)
+{
+  if(msg.poses.size() == 0)
+  {
+	  ROS_ERROR("Received an empty initial pose array.");
+	  return false;
+  }
+
+  boost::recursive_mutex::scoped_lock prl(configuration_mutex_);
+  if(msg.header.frame_id == "")
+  {
+    // This should be removed at some point
+    ROS_WARN("Received initial pose with empty frame_id.  You should always supply a frame_id.");
+  }
+  // We only accept initial pose estimates in the global frame, #5148.
+  // NOTE: Check if this statement is correct as compared to original
+  else if(msg.header.frame_id != amclParams_.global_frame_id_)
+  {
+    ROS_WARN("Ignoring initial pose in frame \"%s\"; initial poses must be in the global frame, \"%s\"",
+             msg.header.frame_id.c_str(),
+             amclParams_.global_frame_id_.c_str());
+    return false;
+  }
+
+  // In case the client sent us a pose estimate in the past, integrate the
+  // intervening odometric change.
+  tf::StampedTransform tx_odom;
+  try
+  {
+    ros::Time now = ros::Time::now();
+    geometry_msgs::TransformStamped t_odom = tf_buffer_.lookupTransform(amclParams_.base_frame_id_, msg.header.stamp,
+                         amclParams_.base_frame_id_, now,
+                         amclParams_.odom_frame_id_, ros::Duration(0.5));
+    tf::transformStampedMsgToTF(t_odom, tx_odom);
```

Comment:
add log lines to see what is happening with the initial pose!! 
put in amcl ros if we need it.

Reply:
tx_odom is identity atleast in simulation! I added a log statement to see check it on chuck